server/src/main/java/org/opensearch/http/AbstractHttpServerTransport.java

```diff
@@ -352,17 +361,30 @@ protected void serverAcceptedChannel(HttpChannel httpChannel) {
      * @param httpChannel that received the http request
      */
     public void incomingRequest(final HttpRequest httpRequest, final HttpChannel httpChannel) {
-        handleIncomingRequest(httpRequest, httpChannel, httpRequest.getInboundException());
+        try (ThreadContext.StoredContext ignore = threadPool.getThreadContext().stashContext()) {
+            final Span span = tracer.startSpan(SpanBuilder.from(httpRequest), httpRequest.getHeaders());
+            try (final SpanScope httpRequestSpanScope = tracer.withSpanInScope(span)) {
+                HttpChannel traceableHttpChannel = new TraceableHttpChannel(httpChannel, span);
+                handleIncomingRequest(httpRequest, traceableHttpChannel, httpRequest.getInboundException());
+            }
+        }
     }
 
     // Visible for testing
     void dispatchRequest(final RestRequest restRequest, final RestChannel channel, final Throwable badRequestCause) {
-        final ThreadContext threadContext = threadPool.getThreadContext();
-        try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-            if (badRequestCause != null) {
-                dispatcher.dispatchBadRequest(channel, threadContext, badRequestCause);
-            } else {
-                dispatcher.dispatchRequest(restRequest, channel, threadContext);
+        RestChannel traceableRestChannel = channel;
+        final Span span = tracer.startSpan(SpanBuilder.from(restRequest));
+        try (final SpanScope spanScope = tracer.withSpanInScope(span)) {
+            if (channel != null) {
+                traceableRestChannel = new TraceableRestChannel(channel, span);
+            }
+            final ThreadContext threadContext = threadPool.getThreadContext();
+            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+                if (badRequestCause != null) {
```

Comment:
Shouldn't the `try (final SpanScope spanScope = tracer.withSpanInScope(span)) {` moved into `try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {` block, so we would propagate the right span into `ThreadContext.StoredContext`?

Reply:
make sense.